zerver/lib/drafts.py

```diff
@@ -1,60 +1,47 @@
 import time
 from functools import wraps
-from typing import Any, Callable, Dict, List, Set
+from typing import Any, Callable, Dict, List, Literal, Union
 
 from django.core.exceptions import ValidationError
 from django.http import HttpRequest, HttpResponse
 from django.utils.translation import gettext as _
-from typing_extensions import Concatenate, ParamSpec
+from pydantic import BaseModel, ConfigDict
+from typing_extensions import Annotated, Concatenate, ParamSpec
 
 from zerver.lib.addressee import get_user_profiles_by_ids
 from zerver.lib.exceptions import JsonableError, ResourceNotFoundError
 from zerver.lib.message import normalize_body, truncate_topic
 from zerver.lib.recipient_users import recipient_for_user_profiles
 from zerver.lib.streams import access_stream_by_id
 from zerver.lib.timestamp import timestamp_to_datetime
-from zerver.lib.validator import (
-    check_dict_only,
-    check_float,
-    check_int,
-    check_list,
-    check_required_string,
-    check_string,
-    check_string_in,
-    check_union,
-)
+from zerver.lib.typed_endpoint import RequiredStringConstraint
 from zerver.models import Draft, UserProfile
 from zerver.tornado.django_api import send_event
 
 ParamT = ParamSpec("ParamT")
-VALID_DRAFT_TYPES: Set[str] = {"", "private", "stream"}
-
-# A validator to verify if the structure (syntax) of a dictionary
-# meets the requirements to be a draft dictionary:
-draft_dict_validator = check_dict_only(
-    required_keys=[
-        ("type", check_string_in(VALID_DRAFT_TYPES)),
-        ("to", check_list(check_int)),  # The ID of the stream to send to, or a list of user IDs.
-        ("topic", check_string),  # This string can simply be empty for private type messages.
-        ("content", check_required_string),
-    ],
-    optional_keys=[
-        ("timestamp", check_union([check_int, check_float])),  # A Unix timestamp.
-    ],
-)
+
+
+class DraftData(BaseModel):
+    model_config = ConfigDict(extra="forbid")
+
+    type: Literal["private", "stream", ""]
+    to: List[int]
+    topic: str
+    content: Annotated[str, RequiredStringConstraint()]
+    timestamp: Union[int, float, None] = None
 
 
 def further_validated_draft_dict(
-    draft_dict: Dict[str, Any], user_profile: UserProfile
+    draft_dict: DraftData, user_profile: UserProfile
 ) -> Dict[str, Any]:
     """Take a draft_dict that was already validated by draft_dict_validator then
     further sanitize, validate, and transform it. Ultimately return this "further
     validated" draft dict. It will have a slightly different set of keys the values
     for which can be used to directly create a Draft object."""
 
-    content = normalize_body(draft_dict["content"])
+    content = normalize_body(draft_dict.content)
 
-    timestamp = draft_dict.get("timestamp", time.time())
+    timestamp = draft_dict.timestamp or time.time()
```

Comment:
This migration may is incorrect in the case that an integer/float timestamp of `0` is passed by the client. Might not matter in this case but we should avoid the pattern.

(I try to avoid using falsy constructions in Python wherever possible in production code because it often has implicit bugs like this)